### PR TITLE
lto: use the `object` crate to load bitcode sections

### LINKED ISSUE #115069: LLVM 18 breakage: `failed to get bitcode from object file` after LLVM 6b539f5eb8e
Filing this to keep track of llvm/llvm-project@6b539f5eb8ef1d3a3c87873caa2dbd5147e1adbd breaking `tests/codegen/lto-removes-invokes.rs` and `tests/codegen/sanitizer/sanitizer-recover.rs#MSAN-RECOVER-LTO`, both with the message `error: failed to get bitcode from object file for LTO (Bitcode section not found in object file)`. 

In https://reviews.llvm.org/D152973#4520240 it sounds like we should already have an idea of what to do here, but I also have the vague notion it might not be a quick fix, so I'm opening a bug that I can refer the handful of Rust-at-HEAD/LLVM-at-HEAD users to until we get this resolved.

### PR DESCRIPTION
**Pull request by durin42**
Upstream change
llvm/llvm-project@6b539f5eb8ef1d3a3c87873caa2dbd5147e1adbd changed `isSectionBitcode` works and it now only respects `.llvm.lto` sections instead of also `.llvmbc`, which it says was never intended to be used for LTO. We already have the object crate handy, so it's easy to use that to load the section.

We have to sniff for raw bitcode files by hand, as `object` doesn't support those, but that's easy enough. I also look for both .llvmbc and .llvm.lto sections so that the behavior of the new code matches the old LLVM behavior exactly, though .llvm.lto sections seem to not (currently) be covered by tests.

Fixes #115069

r? @nikic
@rustbot label: +llvm-main

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/back/lto.rs
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/back/lto.rs:154]: The section name is platform dependent, see https://github.com/rust-lang/rust/blob/97fff1f2ed01f6f7c0c204530b693c74d88c2105/compiler/rustc_codegen_llvm/src/back/write.rs#L935.

Comment (durin42) [reply compiler/rustc_codegen_llvm/src/back/lto.rs]: TIL. Fixed.

File: compiler/rustc_codegen_llvm/src/back/lto.rs
Comment (durin42) [reply compiler/rustc_codegen_llvm/src/back/lto.rs]: Whoops. Fixed.

File: compiler/rustc_codegen_llvm/src/back/lto.rs
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/back/lto.rs:158]: Can we instead factor out a function that returns the correct section name and then look that up? That way things will stay in sync.

File: compiler/rustc_codegen_llvm/src/back/write.rs
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/back/write.rs:856]: I think it would be a bit cleaner to accept cgcx in this function and make it do the is_apple / is_aix check itself, so we don't have to pass around the two separate parameters.

Comment (durin42) [reply compiler/rustc_codegen_llvm/src/back/write.rs]: sure

**Review by nikic — APPROVED** [review]

**Comment by durin42** [comment]
@nikic do you need anything else from me? I see you didn't tell bors to merge this.

**Comment by durin42** [comment]
@nikic back to you, I'm afraid: we had to enable the `wasm` feature in `object` otherwise wasm targets end up broken.

Sigh.

**Comment by durin42** [comment]
Sigh. @rust-lang/compiler are we okay adding `wasmparser` as a dep of rustc? The list says to ask the team before expanding the list.

**Comment by bjorn3** [comment]
The wasm support in object only supports linked wasm modules, not wasm object files. I presume the reason it works here amyway is because the bitcode would get duplicated between the memory segment (which the object crate doesn't know how to split based on the linkage section) and custom section (which has the same format as wasm modules) or something like that.

**Comment by durin42** [comment]
@bjorn3 should I just go back to calling the LLVM method if the file magic lines up with wasm?

**Comment by bjorn3** [comment]
I think the bitcode will likely stay in a custom section, so using object will work fine. And if it no longer ends up in a custom section in the future you could either switch back or hopefully object supports actually parsing wasm object with the linking custom section files by then.

File: compiler/rustc_codegen_llvm/Cargo.toml
Comment (klensy) [inline compiler/rustc_codegen_llvm/Cargo.toml:17]: Today it seems [not possible](https://github.com/rust-lang/cargo/issues/1197) to gate this behind wasm target, to not to pull this on other platforms?

**Comment by durin42** [comment]
Yes, as far as I know I'm waiting on a review.

To my knowledge, there's no way to gate the wasm dep, yes.

File: compiler/rustc_codegen_llvm/Cargo.toml
Comment (bjorn3) [reply compiler/rustc_codegen_llvm/Cargo.toml:17]: It doesn't make sense to do that. If we only enable it when rustc itself is compiled for wasm, it wouldn't be possible to cross-compile user code to wasm.

Comment (klensy) [reply compiler/rustc_codegen_llvm/Cargo.toml:17]: Ok.

**Comment by nikic** [comment]
r=me on the code change, but I believe the license needs to go under EXCEPTIONS, because it doesn't comply with Rust's license policy, so it's only allowed by dint of being a rustc only dep. Would be great if someone (@wesleywiser?) can confirm the license situation here.

**Comment by durin42** [comment]
@wesleywiser do you need anything from me on the license front here?

**Comment by zmodem** [comment]
We just tried cherry-picking this in Chromium, but it seems that on Mac this actually breaks tests/codegen/lto-removes-invokes.rs rather than fixing it (https://crbug.com/1476119#c3). Is there some CI that runs on Mac that can be triggered here?

**Comment by davidtwco** [comment]
> We just tried cherry-picking this in Chromium, but it seems that on Mac this actually breaks tests/codegen/lto-removes-invokes.rs rather than fixing it (https://crbug.com/1476119#c3). Is there some CI that runs on Mac that can be triggered here?

CI on macOS will be run during the merge testing so if there's a failure there then it'll be fixed before this gets merged.

**Comment by davidtwco** [comment]
> r=me on the code change, but I believe the license needs to go under EXCEPTIONS, because it doesn't comply with Rust's license policy, so it's only allowed by dint of being a rustc only dep. Would be great if someone (@wesleywiser?) can confirm the license situation here.

I think this is okay because the dependency is _Apache 2.0 with LLVM Exception_ licensed, which is compatible with rustc's license.

**Comment by durin42** [comment]
Obsoleted by #115641

**Pull request closed without merging**
